internal/internal_workflow_client.go

```diff
@@ -1170,3 +950,273 @@ func serializeSearchAttributes(input map[string]interface{}) (*commonpb.SearchAt
+		Header:                   header,
+	}
+
+	var response *workflowservice.SignalWithStartWorkflowExecutionResponse
+
+	// Start creating workflow request.
+	grpcCtx, cancel := newGRPCContext(ctx, defaultGrpcRetryParameters(ctx))
+	defer cancel()
+
+	response, err = w.client.workflowService.SignalWithStartWorkflowExecution(grpcCtx, signalWithStartRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	iterFn := func(fnCtx context.Context, fnRunID string) HistoryEventIterator {
+		rpcScope := metrics.GetMetricsScopeForRPC(w.client.metricsScope, in.WorkflowType,
+			metrics.NoneTagValue, in.Options.TaskQueue)
+		return w.client.getWorkflowHistory(fnCtx, in.Options.ID, fnRunID, true,
+			enumspb.HISTORY_EVENT_FILTER_TYPE_CLOSE_EVENT, rpcScope)
+	}
+
+	curRunIDCell := util.PopulatedOnceCell(response.GetRunId())
+	return &workflowRunImpl{
+		workflowType:  in.WorkflowType,
+		workflowID:    in.Options.ID,
+		firstRunID:    response.GetRunId(),
+		currentRunID:  &curRunIDCell,
+		iterFn:        iterFn,
+		dataConverter: w.client.dataConverter,
+		registry:      w.client.registry,
+	}, nil
+}
+
+func (w *workflowClientInterceptor) CancelWorkflow(ctx context.Context, in *ClientCancelWorkflowInput) error {
+	request := &workflowservice.RequestCancelWorkflowExecutionRequest{
+		Namespace: w.client.namespace,
+		WorkflowExecution: &commonpb.WorkflowExecution{
+			WorkflowId: in.WorkflowID,
+			RunId:      in.RunID,
+		},
+		Identity: w.client.identity,
+	}
+	grpcCtx, cancel := newGRPCContext(ctx, defaultGrpcRetryParameters(ctx))
+	defer cancel()
+	_, err := w.client.workflowService.RequestCancelWorkflowExecution(grpcCtx, request)
+	return err
+}
+
+func (w *workflowClientInterceptor) TerminateWorkflow(ctx context.Context, in *ClientTerminateWorkflowInput) error {
+	datailsPayload, err := w.client.dataConverter.ToPayloads(in.Details...)
+	if err != nil {
+		return err
+	}
+
+	request := &workflowservice.TerminateWorkflowExecutionRequest{
+		Namespace: w.client.namespace,
+		WorkflowExecution: &commonpb.WorkflowExecution{
+			WorkflowId: in.WorkflowID,
+			RunId:      in.RunID,
+		},
+		Reason:   in.Reason,
+		Identity: w.client.identity,
+		Details:  datailsPayload,
+	}
+
+	grpcCtx, cancel := newGRPCContext(ctx, defaultGrpcRetryParameters(ctx))
+	defer cancel()
+	_, err = w.client.workflowService.TerminateWorkflowExecution(grpcCtx, request)
+	return err
+}
+
+func (w *workflowClientInterceptor) QueryWorkflow(
+	ctx context.Context,
+	in *ClientQueryWorkflowInput,
+) (converter.EncodedValue, error) {
+	result, err := w.client.QueryWorkflowWithOptions(ctx, &QueryWorkflowWithOptionsRequest{
+		WorkflowID: in.WorkflowID,
+		RunID:      in.RunID,
+		QueryType:  in.QueryType,
+		Args:       in.Args,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return result.QueryResult, nil
+}
+
+// Required to implement ClientOutboundInterceptor
+func (*workflowClientInterceptor) mustEmbedClientOutboundInterceptorBase() {}
```

Comment:
Where is ResetWorkflowExecution?

Reply:
There are many client calls that are not intercepted. From https://github.com/temporalio/proposals/pull/45:

> We intentionally choose to only allow a subset of client calls to be mocked as these are the ones considered "workflow client" calls by the SDKs

Every other SDK has only a few high-level client calls in what they call a "workflow client" and then asks the user to use gRPC for the rest. But Go is different in that it hides gRPC adds many other high-level client abstractions. It was decided for consistency reasons to only intercept the client calls that all SDKs allow to be intercepted instead of intercepting every client call.